src/driver/driver.hpp

```diff
@@ -83,25 +82,55 @@ namespace DriverUtils {
 
 template <typename T, class... Args>
 TaskListStatus ConstructAndExecuteBlockTasks(T *driver, Args... args) {
+#ifdef OPENMP_PARALLEL
+  int nthreads = driver->pmesh->GetNumMeshThreads();
+#endif
   int nmb = driver->pmesh->GetNumMeshBlocksThisRank(Globals::my_rank);
   std::vector<TaskList> task_lists;
   MeshBlock *pmb = driver->pmesh->pblock;
+  int i = 0;
   while (pmb != nullptr) {
     task_lists.push_back(driver->MakeTaskList(pmb, std::forward<Args>(args)...));
+    // reassigning ExecSpaces to account for changing MeshBlocks due to, e.g.,
+    // loadbalance or AMR
+    pmb->exec_space = driver->pmesh->GetExecSpaceFromPool(i);
     pmb = pmb->next;
+    i += 1;
   }
   int complete_cnt = 0;
-  while (complete_cnt != nmb) {
-    // TODO(pgrete): need to let Kokkos::PartitionManager handle this
-    for (auto i = 0; i < nmb; ++i) {
-      if (!task_lists[i].IsComplete()) {
-        auto status = task_lists[i].DoAvailable();
-        if (status == TaskListStatus::complete) {
-          complete_cnt++;
+  Kokkos::View<int *, HostMemSpace> mb_locks("mb_locks", nmb);
+  auto f = [&](int thread_id, int num_threads) {
+    while (complete_cnt != nmb) {
+      for (auto i = 0; i < nmb; ++i) {
+        // Workaround to ensure that the same thread works on the same MeshBlocks within
+        // a cycle. Trying to circumvent problem in setting scratch pad memory, which is
+        // not thread safe.
+        if (i % num_threads != thread_id) {
+          continue;
+        }
+        // try to obtain lock by changing val from 0 to 1
+        if (Kokkos::atomic_compare_exchange_strong(&mb_locks(i), 0, 1)) {
```

Comment:
Is this necessary given the previous check? I would think if only one thread can ever operate on a single mesh block, you shouldn't need to lock access to associated tasks.

Reply:
Good point. This is a remnant of the first iteration of that block. I'll remove it.